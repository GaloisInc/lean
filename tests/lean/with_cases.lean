open tactic

example (p q : Prop) [s₁ : decidable p] [s₂ : decidable q] : true :=
begin
  unfreeze_local_instances,
  with_cases { cases s₁; cases s₂ },
  trace_state,
  all_goals { intros, trivial }
end

def split (p : Prop) {q : Prop} [decidable p] (pos : p → q) (neg : ¬ p → q) : q :=
decidable.by_cases pos neg

example (p q : Prop) [decidable p] [decidable q] : true :=
begin
  with_cases { apply split p; apply split q; intros hq hp },
  trace_state,
  all_goals { intros, trivial }
end
