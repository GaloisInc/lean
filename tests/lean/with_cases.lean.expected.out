4 goals
case decidable.is_false, decidable.is_false
p q : Prop
⊢ ¬p → ¬q → true

case decidable.is_false, decidable.is_true
p q : Prop
⊢ ¬p → q → true

case decidable.is_true, decidable.is_false
p q : Prop
⊢ p → ¬q → true

case decidable.is_true, decidable.is_true
p q : Prop
⊢ p → q → true
4 goals
case pos, pos
p q : Prop,
_inst_1 : decidable p,
_inst_2 : decidable q
⊢ q → p → true

case pos, neg
p q : Prop,
_inst_1 : decidable p,
_inst_2 : decidable q
⊢ ¬q → p → true

case neg, pos
p q : Prop,
_inst_1 : decidable p,
_inst_2 : decidable q
⊢ q → ¬p → true

case neg, neg
p q : Prop,
_inst_1 : decidable p,
_inst_2 : decidable q
⊢ ¬q → ¬p → true
